File: src/main/kotlin/magicbook/gtlitecore/api/GTLiteAPI.java
package magicbook.gtlitecore.api;

import gregtech.api.GregTechAPI;
import gregtech.api.creativetab.BaseCreativeTab;
import gregtech.common.blocks.BlockCleanroomCasing;
import gregtech.common.blocks.BlockFusionCasing;
import gregtech.common.blocks.BlockGlassCasing;
import gregtech.common.blocks.MetaBlocks;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import magicbook.gtlitecore.api.advancement.IAdvancementManager;
import magicbook.gtlitecore.api.block.IBlockTier;
import magicbook.gtlitecore.api.block.impl.WrappedIntTier;
import magicbook.gtlitecore.api.command.ICommandManager;
import magicbook.gtlitecore.api.module.IModuleManager;
import magicbook.gtlitecore.api.network.INetworkHandler;
import magicbook.gtlitecore.api.sound.ISoundManager;
import magicbook.gtlitecore.common.block.GTLiteMetaBlocks;
import magicbook.gtlitecore.common.block.blocks.BlockComponentAssemblyCasing;
import magicbook.gtlitecore.common.block.blocks.BlockConveyorCasing;
import magicbook.gtlitecore.common.block.blocks.BlockEmitterCasing;
import magicbook.gtlitecore.common.block.blocks.BlockFieldGenCasing;
import magicbook.gtlitecore.common.block.blocks.BlockFusionCasing01;
import magicbook.gtlitecore.common.block.blocks.BlockFusionCasing02;
import magicbook.gtlitecore.common.block.blocks.BlockFusionCasing03;
import magicbook.gtlitecore.common.block.blocks.BlockGlassCasing02;
import magicbook.gtlitecore.common.block.blocks.BlockMotorCasing;
import magicbook.gtlitecore.common.block.blocks.BlockPistonCasing;
import magicbook.gtlitecore.common.block.blocks.BlockProcessorCasing;
import magicbook.gtlitecore.common.block.blocks.BlockPumpCasing;
import magicbook.gtlitecore.common.block.blocks.BlockRobotArmCasing;
import magicbook.gtlitecore.common.block.blocks.BlockSensorCasing;
import magicbook.gtlitecore.common.block.blocks.BlockWireCoils;
import magicbook.gtlitecore.common.item.GTLiteMetaItems;
import net.minecraft.block.state.IBlockState;

public class GTLiteAPI
{
    /* -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ API Objects +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- */
    // Will always be available.
    public static Object instance;
    // Will be available at the Construction stage.
    public static IModuleManager moduleManager;
    // Will be available at the Pre-Initialization stage
    public static INetworkHandler networkHandler;
    // Will be available at the Server-Starting stage.
    public static ICommandManager commandManager;
    // Will be available at the Pre-Initialization stage.
    public static IAdvancementManager advancementManager;
    // Will be available at the Pre-Initialization stage.
    public static ISoundManager soundManager;

    /* -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- Creative Tabs -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- */
    // Total CreativeTabs of gtlitecore, this is default settings for many contents in
    // the mod, but if item, block or other contents satisfied descriptions of the
    // following CreativeTabs, then should use consistent CreativeTabs.
    public static final BaseCreativeTab TAB_GTLITE = new BaseCreativeTab("gtlite",
            () -> GTLiteMetaItems.LOGO_CORE.getStackForm(), false);

    // Machines (consists of single machines and multiblock machines) for gtlitecore,
    // which copied from gregtech's machine creative tabs.
    public static final BaseCreativeTab TAB_GTLITE_MACHINE = new BaseCreativeTab("gtlite.machine",
            () -> GTLiteMetaItems.LOGO_MACHINE.getStackForm(), false);

    // Decorations for gtlitecore which consists of all stone blocks and its variants,
    // trees, planks, logs and other tree related blocks, and decorative blocks.
    public static final BaseCreativeTab TAB_GTLITE_DECORATION = new BaseCreativeTab("gtlite.decoration",
            () -> GTLiteMetaItems.LOGO_DECORATION.getStackForm(), false);

    /* -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ BlockState Maps +-+-+-+-+-+-+-+-+-+-+-+-+-+-+- */
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_MOTOR_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_PISTON_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_PUMP_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_CONVEYOR_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_ROBOT_ARM_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_EMITTER_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_SENSOR_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_FIELD_GEN_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_PROCESSOR_CASING = new Object2ObjectOpenHashMap<>();

    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_BOROSILICATE_GLASS = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_FUSION_CASING = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_FUSION_COIL = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_CRYOSTAT = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_DIVERTOR = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_VACUUM = new Object2ObjectOpenHashMap<>();
    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_COMPONENT_CASING = new Object2ObjectOpenHashMap<>();

    public static final Object2ObjectOpenHashMap<IBlockState, IBlockTier> MAP_CLEANROOM_CASING = new Object2ObjectOpenHashMap<>();
    /* -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- */
    public static void init()
    {
        // motorCasings
        for (BlockMotorCasing.MotorCasingTier tier : BlockMotorCasing.MotorCasingTier.values())
        {
            MAP_MOTOR_CASING.put(GTLiteMetaBlocks.MOTOR_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // pistonCasings
        for (BlockPistonCasing.PistonCasingTier tier : BlockPistonCasing.PistonCasingTier.values())
        {
            MAP_PISTON_CASING.put(GTLiteMetaBlocks.PISTON_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // pumpCasings
        for (BlockPumpCasing.PumpCasingTier tier : BlockPumpCasing.PumpCasingTier.values())
        {
            MAP_PUMP_CASING.put(GTLiteMetaBlocks.PUMP_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // conveyorCasings
        for (BlockConveyorCasing.ConveyorCasingTier tier : BlockConveyorCasing.ConveyorCasingTier.values())
        {
            MAP_CONVEYOR_CASING.put(GTLiteMetaBlocks.CONVEYOR_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // robotArmCasings
        for (BlockRobotArmCasing.RobotArmCasingTier tier : BlockRobotArmCasing.RobotArmCasingTier.values())
        {
            MAP_ROBOT_ARM_CASING.put(GTLiteMetaBlocks.ROBOT_ARM_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // emitterCasings
        for (BlockEmitterCasing.EmitterCasingTier tier : BlockEmitterCasing.EmitterCasingTier.values())
        {
            MAP_EMITTER_CASING.put(GTLiteMetaBlocks.EMITTER_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // sensorCasings
        for (BlockSensorCasing.SensorCasingTier tier : BlockSensorCasing.SensorCasingTier.values())
        {
            MAP_SENSOR_CASING.put(GTLiteMetaBlocks.SENSOR_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // fieldGenCasings
        for (BlockFieldGenCasing.FieldGenCasingTier tier : BlockFieldGenCasing.FieldGenCasingTier.values())
        {
            MAP_FIELD_GEN_CASING.put(GTLiteMetaBlocks.FIELD_GEN_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // processorCasings
        for (BlockProcessorCasing.ProcessorCasingTier tier : BlockProcessorCasing.ProcessorCasingTier.values())
        {
            MAP_PROCESSOR_CASING.put(GTLiteMetaBlocks.PROCESSOR_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // borosilicateGlasses
        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.BOROSILICATE, 1));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.TITANIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.TITANIUM_BOROSILICATE, 2));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.TUNGSTEN_STEEL_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.TUNGSTEN_STEEL_BOROSILICATE, 3));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.RHODIUM_PLATED_PALLADIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.RHODIUM_PLATED_PALLADIUM_BOROSILICATE, 4));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.OSMIRIDIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.OSMIRIDIUM_BOROSILICATE, 5));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.TRITANIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.TRITANIUM_BOROSILICATE, 6));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.NEUTRONIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.NEUTRONIUM_BOROSILICATE, 7));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.COSMIC_NEUTRONIUM_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.COSMIC_NEUTRONIUM_BOROSILICATE, 8));

        MAP_BOROSILICATE_GLASS.put(GTLiteMetaBlocks.TRANSPARENT_CASING_02.getState(BlockGlassCasing02.GlassType.INFINITY_BOROSILICATE),
                new WrappedIntTier(BlockGlassCasing02.GlassType.INFINITY_BOROSILICATE, 9));

        // TODO UXV+ Borosilicates.

        // fusion related casings (fusion casing, coil, cryostat, divertor and vacuum)
        MAP_FUSION_CASING.put(MetaBlocks.FUSION_CASING.getState(BlockFusionCasing.CasingType.FUSION_CASING),
                new WrappedIntTier(BlockFusionCasing.CasingType.FUSION_CASING, 1));

        MAP_FUSION_CASING.put(MetaBlocks.FUSION_CASING.getState(BlockFusionCasing.CasingType.FUSION_CASING_MK2),
                new WrappedIntTier(BlockFusionCasing.CasingType.FUSION_CASING_MK2, 2));

        MAP_FUSION_CASING.put(MetaBlocks.FUSION_CASING.getState(BlockFusionCasing.CasingType.FUSION_CASING_MK3),
                new WrappedIntTier(BlockFusionCasing.CasingType.FUSION_CASING_MK3, 3));

        MAP_FUSION_CASING.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.FUSION_CASING_MK4),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.FUSION_CASING_MK4, 4));

        MAP_FUSION_CASING.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.FUSION_CASING_MK5),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.FUSION_CASING_MK5, 5));

        MAP_FUSION_COIL.put(MetaBlocks.TRANSPARENT_CASING.getState(BlockGlassCasing.CasingType.FUSION_GLASS),
                new WrappedIntTier(BlockGlassCasing.CasingType.FUSION_GLASS, 1));

        MAP_FUSION_COIL.put(MetaBlocks.FUSION_CASING.getState(BlockFusionCasing.CasingType.SUPERCONDUCTOR_COIL),
                new WrappedIntTier(BlockFusionCasing.CasingType.SUPERCONDUCTOR_COIL, 2));

        MAP_FUSION_COIL.put(MetaBlocks.FUSION_CASING.getState(BlockFusionCasing.CasingType.FUSION_COIL),
                new WrappedIntTier(BlockFusionCasing.CasingType.FUSION_COIL, 3));

        MAP_FUSION_COIL.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.ADVANCED_FUSION_COIL),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.ADVANCED_FUSION_COIL, 4));

        MAP_FUSION_COIL.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.ULTIMATE_FUSION_COIL),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.ULTIMATE_FUSION_COIL, 5));

        MAP_CRYOSTAT.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK1),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK1, 1));

        MAP_CRYOSTAT.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK2),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK2, 2));

        MAP_CRYOSTAT.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK3),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK3, 3));

        MAP_CRYOSTAT.put(GTLiteMetaBlocks.FUSION_CASING_01.getState(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK4),
                new WrappedIntTier(BlockFusionCasing01.FusionCasingType.CRYOSTAT_MK4, 4));

        MAP_CRYOSTAT.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.CRYOSTAT_MK5),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.CRYOSTAT_MK5, 5));

        MAP_DIVERTOR.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK1),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK1, 1));

        MAP_DIVERTOR.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK2),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK2, 2));

        MAP_DIVERTOR.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK3),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK3, 3));

        MAP_DIVERTOR.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK4),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK4, 4));

        MAP_DIVERTOR.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK5),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.DIVERTOR_MK5, 5));

        MAP_VACUUM.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.VACUUM_MK1),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.VACUUM_MK1, 1));

        MAP_VACUUM.put(GTLiteMetaBlocks.FUSION_CASING_02.getState(BlockFusionCasing02.FusionCasingType.VACUUM_MK2),
                new WrappedIntTier(BlockFusionCasing02.FusionCasingType.VACUUM_MK2, 2));

        MAP_VACUUM.put(GTLiteMetaBlocks.FUSION_CASING_03.getState(BlockFusionCasing03.FusionCasingType.VACUUM_MK3),
                new WrappedIntTier(BlockFusionCasing03.FusionCasingType.VACUUM_MK3, 3));

        MAP_VACUUM.put(GTLiteMetaBlocks.FUSION_CASING_03.getState(BlockFusionCasing03.FusionCasingType.VACUUM_MK4),
                new WrappedIntTier(BlockFusionCasing03.FusionCasingType.VACUUM_MK4, 4));

        MAP_VACUUM.put(GTLiteMetaBlocks.FUSION_CASING_03.getState(BlockFusionCasing03.FusionCasingType.VACUUM_MK5),
                new WrappedIntTier(BlockFusionCasing03.FusionCasingType.VACUUM_MK5, 5));

        // componentCasings
        for (BlockComponentAssemblyCasing.ComponentCasingType tier : BlockComponentAssemblyCasing.ComponentCasingType.values())
        {
            MAP_COMPONENT_CASING.put(GTLiteMetaBlocks.COMPONENT_ASSEMBLY_CASING.getState(tier),
                    new WrappedIntTier(tier, tier.ordinal() + 1));
        }

        // cleanroomCasings
        MAP_CLEANROOM_CASING.put(MetaBlocks.CLEANROOM_CASING.getState(BlockCleanroomCasing.CasingType.FILTER_CASING),
                new WrappedIntTier(BlockCleanroomCasing.CasingType.FILTER_CASING, 1));

        MAP_CLEANROOM_CASING.put(MetaBlocks.CLEANROOM_CASING.getState(BlockCleanroomCasing.CasingType.FILTER_CASING_STERILE),
                new WrappedIntTier(BlockCleanroomCasing.CasingType.FILTER_CASING_STERILE, 2));

        // heatingCoils
        for (BlockWireCoils.WireCoilType type : BlockWireCoils.WireCoilType.values())
        {
            GregTechAPI.HEATING_COILS.put(GTLiteMetaBlocks.WIRE_COIL.getState(type), type);
        }

    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/GTLiteAPI.kt
package gregtechlite.gtlitecore.api

import gregtech.api.GregTechAPI
import gregtechlite.gtlitecore.api.advancement.AdvancementManager
import gregtechlite.gtlitecore.api.block.attribute.BlockAttributeRegistry
import gregtechlite.gtlitecore.api.block.attribute.BlockAttributeRegistry.Companion.registerBlockVariants
import gregtechlite.gtlitecore.api.block.attribute.StateTier
import gregtechlite.gtlitecore.api.command.CommandManager
import gregtechlite.gtlitecore.api.module.ModuleManager
import gregtechlite.gtlitecore.api.network.NetworkHandler
import gregtechlite.gtlitecore.api.sound.SoundManager
import gregtechlite.gtlitecore.common.block.adapter.GTCleanroomCasing
import gregtechlite.gtlitecore.common.block.adapter.GTFusionCasing
import gregtechlite.gtlitecore.common.block.adapter.GTGlassCasing
import gregtechlite.gtlitecore.common.block.variant.ComponentAssemblyCasing
import gregtechlite.gtlitecore.common.block.variant.Crucible
import gregtechlite.gtlitecore.common.block.variant.GlassCasing
import gregtechlite.gtlitecore.common.block.variant.Manipulator
import gregtechlite.gtlitecore.common.block.variant.NuclearReactorCore
import gregtechlite.gtlitecore.common.block.variant.ShieldingCore
import gregtechlite.gtlitecore.common.block.variant.WireCoil
import gregtechlite.gtlitecore.common.block.variant.aerospace.AccelerationTrack
import gregtechlite.gtlitecore.common.block.variant.component.ConveyorCasing
import gregtechlite.gtlitecore.common.block.variant.component.EmitterCasing
import gregtechlite.gtlitecore.common.block.variant.component.FieldGenCasing
import gregtechlite.gtlitecore.common.block.variant.component.MotorCasing
import gregtechlite.gtlitecore.common.block.variant.component.PistonCasing
import gregtechlite.gtlitecore.common.block.variant.component.ProcessorCasing
import gregtechlite.gtlitecore.common.block.variant.component.PumpCasing
import gregtechlite.gtlitecore.common.block.variant.component.RobotArmCasing
import gregtechlite.gtlitecore.common.block.variant.component.SensorCasing
import gregtechlite.gtlitecore.common.block.variant.fusion.FusionCasing
import gregtechlite.gtlitecore.common.block.variant.fusion.FusionCoil
import gregtechlite.gtlitecore.common.block.variant.fusion.FusionCryostat
import gregtechlite.gtlitecore.common.block.variant.fusion.FusionDivertor
import gregtechlite.gtlitecore.common.block.variant.fusion.FusionVacuum

object GTLiteAPI
{

    // @formatter:off

    // region API Objects

    // Will always be available.
    lateinit var instance: Any

    // Will be available at the Construction stage.
    lateinit var moduleManager: ModuleManager

    // Will be available at the Pre-Initialization stage
    lateinit var networkHandler: NetworkHandler

    // Will be available at the Server-Starting stage.
    lateinit var commandManager: CommandManager

    // Will be available at the Pre-Initialization stage.
    lateinit var advancementManager: AdvancementManager

    // Will be available at the Pre-Initialization stage.
    lateinit var soundManager: SoundManager

    // endregion

    // region Block Attribute Registries

    @JvmField
    val MOTOR_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("motor_casing_tier")

    @JvmField
    val PISTON_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("piston_casing_tier")

    @JvmField
    val PUMP_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("pump_casing_tier")

    @JvmField
    val CONVEYOR_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("conveyor_casing_tier")

    @JvmField
    val ROBOT_ARM_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("robot_arm_casing_tier")

    @JvmField
    val EMITTER_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("emitter_casing_tier")

    @JvmField
    val SENSOR_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("sensor_casing_tier")

    @JvmField
    val FIELD_GEN_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("field_gen_casing_tier")

    @JvmField
    val PROCESSOR_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("processor_casing_tier")

    @JvmField
    val BOROSILICATE_GLASS_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("borosilicate_glass_tier")

    @JvmField
    val FUSION_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("fusion_casing_tier")

    @JvmField
    val FUSION_COIL_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("fusion_coil_tier")

    @JvmField
    val CRYOSTAT_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("cryostat_tier")

    @JvmField
    val DIVERTOR_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("divertor_tier")

    @JvmField
    val VACUUM_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("vacuum_tier")

    @JvmField
    val COMPONENT_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("component_casing_tier")

    @JvmField
    val NUCLEAR_REACTOR_CORE_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("nuclear_reactor_core_tier")

    @JvmField
    val MANIPULATOR_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("manipulator_tier")

    @JvmField
    val SHIELDING_CORE_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("shielding_core_tier")

    @JvmField
    val ACCELERATION_TRACK_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("acceleration_track_tier")

    @JvmField
    val CLEANROOM_CASING_TIER: BlockAttributeRegistry<Int> = BlockAttributeRegistry.create("cleanroom_casing_tier")

    @JvmField
    val CRUCIBLE_STATS: BlockAttributeRegistry<Crucible> = BlockAttributeRegistry("crucible_stats", StateTier.COMPARATOR)

    // endregion

    fun init()
    {
        MOTOR_CASING_TIER.registerBlockVariants(MotorCasing::class)
        PISTON_CASING_TIER.registerBlockVariants(PistonCasing::class)
        PUMP_CASING_TIER.registerBlockVariants(PumpCasing::class)
        CONVEYOR_CASING_TIER.registerBlockVariants(ConveyorCasing::class)
        ROBOT_ARM_CASING_TIER.registerBlockVariants(RobotArmCasing::class)
        EMITTER_CASING_TIER.registerBlockVariants(EmitterCasing::class)
        SENSOR_CASING_TIER.registerBlockVariants(SensorCasing::class)
        FIELD_GEN_CASING_TIER.registerBlockVariants(FieldGenCasing::class)
        PROCESSOR_CASING_TIER.registerBlockVariants(ProcessorCasing::class)

        BOROSILICATE_GLASS_TIER.registerBlockVariants(GlassCasing.Enum01::class)

        FUSION_CASING_TIER.registerAll(GTFusionCasing.FUSION_CASING.state to 1,
                                       GTFusionCasing.FUSION_CASING_MK2.state to 2,
                                       GTFusionCasing.FUSION_CASING_MK3.state to 3,
                                       FusionCasing.MK4.state to 4,
                                       FusionCasing.MK5.state to 5)

        FUSION_COIL_TIER.registerAll(GTGlassCasing.FUSION_GLASS.state to 1,
                                     GTFusionCasing.SUPERCONDUCTOR_COIL.state to 2,
                                     GTFusionCasing.FUSION_COIL.state to 3,
                                     FusionCoil.ADVANCED.state to 4,
                                     FusionCoil.ULTIMATE.state to 5)

        CRYOSTAT_TIER.registerBlockVariants(FusionCryostat::class)
        DIVERTOR_TIER.registerBlockVariants(FusionDivertor::class)
        VACUUM_TIER.registerBlockVariants(FusionVacuum::class)

        COMPONENT_CASING_TIER.registerBlockVariants(ComponentAssemblyCasing::class)

        NUCLEAR_REACTOR_CORE_TIER.registerBlockVariants(NuclearReactorCore.Enum01::class)
        NUCLEAR_REACTOR_CORE_TIER.registerBlockVariants(NuclearReactorCore.Enum02::class)

        MANIPULATOR_TIER.registerBlockVariants(Manipulator::class)
        SHIELDING_CORE_TIER.registerBlockVariants(ShieldingCore::class)

        ACCELERATION_TRACK_TIER.registerBlockVariants(AccelerationTrack::class)

        CRUCIBLE_STATS.registerBlockVariants(Crucible::class)

        CLEANROOM_CASING_TIER.registerAll(GTCleanroomCasing.FILTER_CASING.state to 1,
                                          GTCleanroomCasing.FILTER_CASING_STERILE.state to 2)

        for (type in WireCoil.entries)
        {
            GregTechAPI.HEATING_COILS.put(type.state, type)
        }

    }

    // @formatter:on

}
